# ======== Packages required =========
if("Xmisc" %in% rownames(installed.packages()) == FALSE) {
  install.packages(Xmisc)}
library(Xmisc)

# === setting environment ===
parser <- ArgumentParser$new()
parser$add_usage('Rsubread_index_cl.R [options]')
parser$add_description('An executable R script parsing arguments from Unix-like command line.')
parser$add_argument('--h',type='logical', action='store_true', help='Print the help page')
parser$add_argument('--help',type='logical',action='store_true',help='Print the help page')
parser$add_argument('--dir', type = 'character', default = getwd(), help = '"directory",Enter your working directory')
# index for reference sequences
parser$add_argument('--i', type = 'character', default = 'hg', help = 'The basename of index file. Index files should be locatedin the current directory')
# input reads and output
parser$add_argument('--t', type = 'character', default = 'rna', help = '"Type",Enter your sample type: "rna" or "dna"')
parser$add_argument('--iF', type = 'character', default = 'gzFASTQ', help = '"input_format",specifying format of read input files:"gzFASTQ","SAM" or"BAM"')
parser$add_argument('--oF', type = 'character', default = 'BAM', help = '"Output_format",specifying format of read output files:"SAM" or"BAM"')
parser$add_argument('--sc',type='logical', default = FALSE ,help='if  the argument to  the readfile1 option  gives  a  directory containing many BCL files or gzipped BCL files directly generated by the sequenser. This option is only for single-cell RNA-seq read mapping.')
# offset value added to Phred quality scores of read bases
parser$add_argument('--ph', type = 'numeric', default = 33, help = 'offset value added to Phred quality scores of read bases')
# thresholds for mapping
parser$add_argument('--ns', type = 'numeric', default = 10, help = 'numeric value giving the number of subreads extracted from each read.')
parser$add_argument('--t1', type = 'numeric', default = 3, help = 'numeric value giving the consensus threshold for reporting a hit.   This is thethreshold for the first reads if paired-end read data are provided.')
parser$add_argument('--t2', type = 'numeric', default = 1, help = 'numeric value giving the consensus threhold for the second reads in paired-enddata.')
parser$add_argument('--mM', type = 'numeric', default = 3, help = 'numeric value giving the maximum number of mis-matched bases allowed inthe alignment.')
# unique mapping and multi-mapping
parser$add_argument('--uq',type='logical', default = FALSE ,help='if only uniquely mapped reads should be reported. A uniquely mapped read has one single mapping location that has less mis-matched basesthan  any  other  candidate  locations')
parser$add_argument('--nB', type = 'numeric', default = 1, help = 'numeric value specifying the maximal number of equally-best mapping loca-tions that will be reported for a multi-mapping read.')
# indel detection
parser$add_argument('--id', type = 'numeric', default = 5, help = 'numeric value giving the maximum number of insertions/deletions allowed during the mapping.')
parser$add_argument('--cI',type='logical', default = FALSE , help='if complex indels will be detected. If TRUE, the program will try to detect multiple short indels that occurs concurrently in a small genomicregion (indels could be as close as 1bp apart)')
# reading trimming
parser$add_argument('--t5', type = 'numeric', default = 0, help = 'numeric value giving the number of bases trimmed off from 5 end of each read.0by default.')
parser$add_argument('--t3', type = 'numeric', default = 0, help = 'numeric value giving the number of bases trimmed off from 3 end of each read.0by default.')
# distance and orientation of paired end reads
parser$add_argument('--miF', type = 'numeric', default = 50, help = 'numeric value giving the minimum fragment length.')
parser$add_argument('--maF', type = 'numeric', default = 600, help = 'numeric value giving the maximum fragment length.')
parser$add_argument('--or', type = 'character', default = 'fr', help = 'giving the orientation of the two reads from the same pair.  It has three possible values including "fr","ff" and "rf"')
# number of CPU threads
parser$add_argument('--Ct', type = 'numeric', default = 1, help = 'numeric value giving the number of threads used for mapping.')
# read group
parser$add_argument('--rGI', type = 'character', default = 'NULL', help = 'a character string giving the read group ID.')
parser$add_argument('--rG', type = 'character', default = 'NULL', help = 'This string will be added to theread group (RG) header in the mapping output.')
# read order
parser$add_argument('--kRO',type='logical', default = FALSE , help='if the order of reads in the BAM output is kept the same as that in the input file. ')
parser$add_argument('--sRC',type='logical', default = FALSE , help='if reads will be sorted by their mapping locations in the map-ping output.  This option is applicable for BAM output only.')
# color space reads
parser$add_argument('--c2b',type='logical', default = TRUE , help='If TRUE, color-space read bases will be converted to base-space basesin the mapping output')
# dynamic programming
parser$add_argument('--GOP', type = 'numeric', default = -1, help = 'a numeric value giving the penalty for opening a gap')
parser$add_argument('--GEP', type = 'numeric', default = 0, help = 'a numeric value giving the penalty for extending the gap')
parser$add_argument('--MP', type = 'numeric', default = 0, help = 'a numeric value giving the penalty for mismatches')
parser$add_argument('--MS', type = 'numeric', default = 2, help = 'a numeric value giving the score for matches')
# detect structural variants
parser$add_argument('--dS',type='logical', default = FALSE , help='if structural variants (SVs) will be detected during read mapping.')
# gene annotation
parser$add_argument('--uA',type='logical', default = FALSE , help='"useAnnotation", if gene annotation information will be used in the mapping')
parser$add_argument('--ai', type = 'character', default = "hg38", help = 'a character string specifying an in-built annotation used for read summarization. It has four possible values including"mm10","mm9","hg38"and"hg19"')
parser$add_argument('--ae', type = 'character', default = 'NULL', help = 'A character string giving name of a user-provided annotation file or a data frame including user-provided annotation data. If the annotation is in GTF format, it can only be provided as a file')
parser$add_argument('--iG',type='logical', default = FALSE, help='"isGTF", if the annotation provided via the annot.extargument is in GTF format or  not')
parser$add_argument('--Gf', type = 'character', default = "exon" , help = 'a character string denoting the type of features that will be extracted from a GTF annotation. This argument is only applicable when --iG==TRUE')
parser$add_argument('--GT', type = 'character', default = "gene_id", help = 'a character string denoting the type of attributes in a GTF annotation that will beused to group features.This argument is only applicable when --iG==TRUE')
parser$add_argument('--cA', type = 'character', default = 'NULL', help = 'the name of a comma-delimited text file that includes aliases of chromosome names')
parser$helpme()
# === variables ====
dirPath <- dir
dirPath <-gsub ('\\\\','/',dirPath)
if (dir.exists(dirPath)){
  print(paste0("Setting ",dirPath," as the working directory"))
} else {
  print("Directory is not existing!")
  quit()
}
index<-i
# Assign R1 to readfile1; R2 to readfile2
fastq_files <- list.files(dirPath,pattern = ".gz$")
r1<-NULL
r2<-NULL
for (f in fastq_files){
  if (grepl('R1', f, fixed = TRUE)){
    r1<-c(r1,f)
  } else if (grepl('R2', f, fixed = TRUE)){
    r2<-c(r2,f)
  }
}
readfile1 <- file.path(dirPath, r1)
readfile2 <- file.path(dirPath, r2)

if (rGI=='NULL'){rGI <-NULL}
if (rG=='NULL'){rG <-NULL}
if (ae=='NULL'){ae <-NULL}
if (cA=='NULL'){cA <-NULL}

# ======== Packages required (cont.) =========
if (!check.packages(Rsubread)){
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")  
  BiocManager::install("Rsubread")}  

library(Rsubread)
# === execute align command

align(
  # index for reference sequences
  index, # 'character', default = 'hg'
  # input reads and output
  readfile1,
  readfile2,
  type = t, # 'character', default = 'rna'
  input_format <- iF, # 'character', default = 'gzFASTQ'
  output_format <- oF, #'character', default = 'BAM'
  output_file = paste(sub("_R1.fastq.gz", "", r1),"align",output_format,sep="."),
  isBCLinput = sc, #'logical', default = FALSE
  # offset value added to Phred quality scores of read bases
  phredOffset = ph, # 'numeric', default = 33
  # thresholds for mapping
  nsubreads = ns, #'numeric', default = 10
  TH1 = t1, # 'numeric', default = 3
  TH2 = t2, # 'numeric', default = 1
  maxMismatches = mM, # 'numeric', default = 3
  # unique mapping and multi-mapping
  unique = uq, #'logical', default = FALSE
  nBestLocations = nB, # 'numeric', default = 1
  # indel detection
  indels = id, # 'numeric', default = 5
  complexIndels = cI, # 'logical', default = FALSE
  # reading trimming
  nTrim5= t5, # 'numeric', default = 0
  nTrim3 = t3, # 'numeric', default = 0
  # distance and orientation of paired end reads
  minFragLength = miF, # 'numeric', default = 50
  maxFragLength = maF, # 'numeric', default = 600
  PE_orientation = or, # character', default = 'fr'
  # number of CPU threads
  nthreads = Ct, # 'numeric', default = 1
  # read group
  readGroupID = rGI,
  readGroup = rG,
  # read order 
  keepReadOrder = kRO,
  sortReadsByCoordinates = sRC,
  # color space reads,
  color2base = c2b,
  # dynamic programming
  DP_GapOpenPenalty = GOP,
  DP_GapExtPenalty = GEP,
  DP_MismatchPenalty = MP,
  DP_MatchScore = MS,
  # detect structural variants
  detectSV = dS,
  # gene annotation
  useAnnotation = uA,
  annot.inbuilt = ai, # c("mm10","mm9","hg38","hg19")
  annot.ext = ae, # gene.gtf
  isGTF = iG,
  GTF.featureType = Gf,
  GTF.attrType = GT,
  chrAliases = cA
)

